storage/types.go

```diff
@@ -79,16 +80,39 @@ func NewNodeIDWithPrefix(prefix uint64, prefixLenBits, nodeIDLenBits, maxLenBits
 	return p
 }
 
+func bitLen(x int64) int {
+	r := 0
+	for x > 0 {
+		r++
+		x >>= 1
+	}
+	return r
+}
+
 // NewNodeIDForTreeCoords creates a new NodeID for a Tree node with a specified depth and
-// index
-func NewNodeIDForTreeCoords(depth int64, index int64, maxLenBits int) NodeID {
-	r := NewEmptyNodeID(maxLenBits)
-	for i := len(r.Path) - 1; index > 0; i-- {
+// index.
+// This method is used exclusively by the Log, and since the Log model grows upwards from the
+// leaves we modify the provided coords accordingly.
+//
+// index is the horizontal index into the tree at level depth, so the returned
+// NodeID will be zero padded on the right by depth places.
+func NewNodeIDForTreeCoords(depth int64, index int64, maxTreeDepth int) (NodeID, error) {
+	bl := bitLen(index)
+	if depth < 0 || bl > int(maxTreeDepth-int(depth)) {
```

Comment:
Yeah, this is a bit confusing :)
So: 
- there can't be `2^64` leaves in here because we use `int64` (mainly so that we *could* easily detect arithmetic truncation - though I suspect we need to add a bunch more checks for that), so we'd have a bad day once we got to `2^63-1`.
- in the 'Trillian log' world, depth does start at 0 from the leaves, so `index` at `depth=63` can only be 1 bit (because it's either the left or right child of the root), whereas a leaf at `depth=0, index=0`, would really have a `nodeID` of `0x0000000000000000` (assuming a 64 level tree)

the check you mention is trying to enforce this 2nd point - I've added a couple more test cases to hopefully illustrate a bit better.

I've also bumped `log.maxTreeDepth` to 63 :)